modules/web/src/main/scala/Gatekeeper.scala

```diff
@@ -79,40 +82,50 @@ object Gatekeeper {
+  def login(env: Environment): HttpService[IO] =
+    HttpService[IO] {
       // curl -i -d '{ "uid": "bobdole", "pass": "banana" }' localhost:8080/login
       case req @ POST -> Root / "login" =>
-        // this is gross, hopefully will improve before 0.18 ships
-        req.as(implicitly, jsonOf[IO, LoginRequest]).flatMap { case LoginRequest(u, p) =>
+        req.decodeJson[LoginRequest].flatMap { case LoginRequest(u, p) =>
           env.tryLogin(u, p).flatMap {
             case None      => Forbidden("Login failed.")
-            case Some(svc) => GemToken.cookie(env, svc.user).flatMap(Ok("Logged in.").addCookie)
+            case Some(svc) => GemToken.cookie(env, svc.user).flatMap(c => Ok("Logged in.").map(_.addCookie(c)))
           }
         }
+    }
 
+  /**
+   * Given an AuthedService and an Environment that knows how to authenticate user cookies, yield a
+   * Normal HttpService that verifies the cookie on the way in and updates it on the way out. This
+   * is a bit more complex than normal services because we must work in OptionT to handle the case
+   * where `delegate` doesn't respond.
+   */
+  def authenticate(env: Environment, delegate: AuthedService[IO, GemService[IO]]): HttpService[IO] =
+    Kleisli[OptionT[IO, ?], Request[IO], Response[IO]] {
       // curl -i -b gem.jwt=... localhost:8080/something/else
       case req =>
-        GemToken.decodeFromCookie(env, req).flatMap {
-          case Left(msg)  => Forbidden(msg)
+        OptionT.liftF(GemToken.decodeFromCookie(env, req)).flatMap {
+          case Left(msg)  => OptionT.liftF(Forbidden(msg))
           case Right(jwt) =>
-            env.service(jwt.uid).flatMap {
-              case None      => Forbidden(s"JWT is valid but user ${jwt.uid} was not found.")
+            OptionT.liftF(env.service(jwt.uid)).flatMap {
+              case None      => OptionT.liftF(Forbidden(s"JWT is valid but user ${jwt.uid} was not found."))
               case Some(svc) =>
-                // Delegate to `authed` and refresh our cookie as the response comes back.
-                authed.run(AuthedRequest(svc, req)).flatMap { res =>
-                  res.cata(r => GemToken.cookie(env, svc.user).map(r.addCookie), Pass.pure[IO])
+                // Delegate and refresh our cookie as the response comes back.
+                delegate.run(AuthedRequest(svc, req)).flatMap { res =>
+                  OptionT.liftF(GemToken.cookie(env, svc.user).map(res.addCookie))
                 }
             }
         }
-
     }
 
+  /**
+   * Construct the gatekeeper middleware. We need a server environment to figure out how to do
+   * do this because we need a way to encode/decode tokens, as well as way to log users in.
+   */
+  @SuppressWarnings(Array("org.wartremover.warts.PublicInference"))
+  def apply(env: Environment): AuthMiddleware[IO, GemService[IO]] = delegate =>
+    login(env) <+> authenticate(env, delegate)
```

Comment:
What does this do?  I can't find `<+>` except in scalaz docs.

Reply:
`<+>` is append for `SemigroupK` (called `Plus` in scalaz), which is like `|+|` for things of kind `* -> *` that ignores the element type. So for `Option[A]` you can say `a |+| b` given `Semigroup[A]` and it smashes the values together if both are `Some`. But `a <+> b` means `a orElse b` … it doesn't require `Semigroup[A]` because it never looks at the elements. 

For `HttpService[IO]`, which is an alias for `Kleisli[OptionT[IO, ?], Request[IO], Response[IO]]` appending `a <+> b` means "send the request to `a` and if it yields `None` then try again with `b`", so it has the same hand-wavey meaning as `SemigroupK[Option]`; i.e., try this and if it doesn't work try the other thing. So this lets you compose (and prioritize) big routes out of smaller ones. Cachai?

Reply:
That used to be done with `|+|` earlier but it was changed for same reason on http4s

Reply:
Yeah the default instance from cats is actually wrong … we're fixing it. https://github.com/typelevel/cats/issues/1928